vllm/lora/ops/xla_ops/pallas.py

```diff
@@ -10,124 +12,487 @@
+        kernel=functools.partial(_bgmv_shrink_kernel, TOKEN_BLOCK, LORA_BLOCK,
+                                 N_LORA_LANES, LANE_SIZE, N),
         out_shape=jax.ShapeDtypeStruct((T, L), dtype=inputs.dtype),
         grid_spec=pltpu.PrefetchScalarGridSpec(
             num_scalar_prefetch=1,
-            grid=(T // TOKENS_BLOCK, L // LORA_RANK_BLOCK,
-                  D // DIM_BLOCK_SIZE),
+            grid=(T // TOKEN_BLOCK, L // LORA_BLOCK, D // DIM_BLOCK),
             in_specs=[
-                pl.BlockSpec((TOKENS_BLOCK, DIM_BLOCK_SIZE),
+                pl.BlockSpec((TOKEN_BLOCK, DIM_BLOCK),
                              lambda i, j, k, block_idx: (i, k)),
-                pl.BlockSpec((N, LORA_RANK_BLOCK, DIM_BLOCK_SIZE),
+                pl.BlockSpec((N, LORA_BLOCK, DIM_BLOCK),
                              lambda i, j, k, block_idx: (0, j, k)),
             ],
-            out_specs=pl.BlockSpec((TOKENS_BLOCK, LORA_RANK_BLOCK),
+            out_specs=pl.BlockSpec((TOKEN_BLOCK, LORA_BLOCK),
                                    lambda i, j, k, block_idx: (i, j)),
             scratch_shapes=[
-                pltpu.VMEM((TOKENS_BLOCK, LORA_RANK_BLOCK), jnp.float32),
-                pltpu.VMEM((TOKENS_BLOCK, LORA_RANK_BLOCK), jnp.float32)
+                pltpu.VMEM((TOKEN_BLOCK, LORA_BLOCK), jnp.float32),
+                pltpu.VMEM((TOKEN_BLOCK, LORA_BLOCK), jnp.float32)
             ]),
         compiler_params=pltpu.TPUCompilerParams(
             dimension_semantics=("parallel", "parallel", "arbitrary")),
         name="bgmv")(idxs, inputs, loras)
 
 
-def bgmv_shape_function(idxs, inputs, loras):
+def bgmv_shrink_shape_function(idxs, inputs, loras):
     T, _ = inputs.shape
     _, L, _ = loras.shape
 
     return [((T, L), inputs.dtype)]
 
 
-XLA_LIB.define("bgmv(Tensor inputs, Tensor loras, Tensor idxs) -> Tensor", )
-
-
-@impl(XLA_LIB, "bgmv", "XLA")
-def bgmv_xla(inputs: torch.Tensor, loras: torch.Tensor, idxs: torch.IntTensor):
+@impl(XLA_LIB, "bgmv_shrink", "XLA")
+def bgmv_shrink_xla(inputs: torch.Tensor, loras: torch.Tensor,
+                    idxs: torch.IntTensor):
     inputs = inputs.to(dtype=loras.dtype)
 
     if len(loras.shape) == 4:
         loras = loras.squeeze(axis=1)
 
+    T, _ = inputs.shape
+    N, L, D = loras.shape
+
+    TOKEN_BLOCK = get_bounded_value(16, next_multiple_of(T, 16), 128)
+    LORA_BLOCK = LORA_RANK_BLOCK_SIZE
+    DIM_BLOCK = largest_divisor(D, [256, 512, 1024])
+
+    # See if we can fit multiple LoRAs in a register. This would activate LoRA
+    # laning
+    N_LORA_LANES = math.ceil(LORA_BLOCK / L)
+    LANE_SIZE = min(L, LORA_BLOCK)
+    if N_LORA_LANES > 1 and N > 1:
+        pad_N = next_multiple_of(N, N_LORA_LANES) - N
+        new_N = N + pad_N
+
+        loras = torch.nn.functional.pad(loras, (0, 0, 0, 0, 0, pad_N))
+        loras = loras.reshape((new_N // N_LORA_LANES, LORA_BLOCK, D))
+        N, L, D = loras.shape
```

Comment:
why are we using capitalized variables?

Reply:
That's a holdover from the first iteration of the kernel when these were constants, can change it